From f403c96dfe77bf9f36b099dcf7c003010a82345a Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Fri, 23 Feb 2024 13:53:56 +0200
Subject: [PATCH] WLANRTSYS-63747 Increase default nl80211 msg size

nlmsg_default_size is pagesize that is 4096 for most platforms.
This is not always enough for the MAX VAPS case.
Therefore it is increased up to 8192 with possibility to increase it
even more with config option nlmsg_default_size.
---
 hostapd/mxl_config.c         | 16 ++++++++++++++++
 src/ap/mxl_config.h          |  1 +
 src/ap/mxl_hostapd.c         | 18 ++++++++++++++++++
 src/drivers/driver_nl80211.c | 26 ++++++++++++++++++++++++++
 4 files changed, 61 insertions(+)

diff --git a/hostapd/mxl_config.c b/hostapd/mxl_config.c
index d20f880d4..70cc10e6c 100644
--- a/hostapd/mxl_config.c
+++ b/hostapd/mxl_config.c
@@ -1795,6 +1795,10 @@ int mxl_hostapd_config_fill (struct hostapd_config *conf,
 				return 0;
 			}
 		}
+	} else if (os_strcmp(buf, "nlmsg_default_size") == 0) {
+		size_t val = (size_t)atoi(pos);
+		if (val > mxl_conf->nlmsg_default_size)
+			mxl_conf->nlmsg_default_size = val;
 	} else
 		return 0;
 
@@ -1865,6 +1869,18 @@ void mxl_hostapd_config_defaults(struct hostapd_config *conf)
 	mxl_conf->basic_eht_mcs_nss_set[0] = set_he_cap(1, EHT_MCS_NSS_FOR_1SS);
 #endif /* CONFIG_IEEE80211BE */
 
+	/*
+	 * nlmsg_default_size is pagesize that is 4096 for most platforms.
+	 * This is not always enough for the MAX VAPS case, therefore increased
+	 * with possibility to increase it even more with config option.
+	 */
+#ifdef CONFIG_AFC
+	/* AFC NL message to driver can exceed 8192 byte */
+	mxl_conf->nlmsg_default_size = 16 * 1024;
+#else
+	mxl_conf->nlmsg_default_size = 8 * 1024;
+#endif
+
 }
 
 void mxl_hostapd_config_free_bss(struct hostapd_bss_config *bss)
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index ad861da79..b627254bf 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -338,6 +338,7 @@ struct mxl_hostapd_config {
 #endif /* CONFIG_IEEE80211BE */
 	int sProbeReqCltMode;
 	int ProbeReqListTimer;
+	size_t nlmsg_default_size;
 };
 
 /**
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 2a3695757..c705ee8c4 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -1806,9 +1806,27 @@ int is_5ghz_op_class(u8 op_class)
 {
 	return op_class >= 115 && op_class <= 130;
 }
+int nl80211_nlmsg_set_default_size_vendor_mxl(size_t nlmsg_new_size);
+
+static void mxl_hostapd_nl80211_driver_init(struct hostapd_iface *iface)
+{
+	size_t nlmsg_default_size = iface->conf->mxl_conf.nlmsg_default_size;
+
+	if (!iface->conf->driver || os_strcmp("nl80211", iface->conf->driver->name)) {
+		wpa_printf(MSG_ERROR, "No nl80211 driver is set for interface");
+		return;
+	}
+
+	if (nl80211_nlmsg_set_default_size_vendor_mxl(nlmsg_default_size)) {
+		wpa_printf(MSG_ERROR, "Failed to set nlmsg_default_size to %zu", nlmsg_default_size);
+	}
+}
+
 void mxl_hostapd_driver_init(struct hostapd_iface *iface, struct wpa_init_params *params)
 {
 	iface->mxl_iface.allow_scan_during_cac = -1;
+
+	mxl_hostapd_nl80211_driver_init(iface);
 }
 
 void mxl_hostapd_setup_interface_pre(struct hostapd_iface *iface)
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 6cc134c11..9f45cbc5b 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -15202,6 +15202,32 @@ wpa_driver_get_multi_hw_info(void *priv, unsigned int *num_multi_hws)
 }
 
 #ifdef CONFIG_VENDOR_MXL
+int nl80211_nlmsg_set_default_size_vendor_mxl(size_t nlmsg_new_size)
+{
+	struct nl_msg *msg;
+	size_t nlmsg_current_size;
+
+	msg = nlmsg_alloc();
+	if (!msg)
+		return -1;
+
+	nlmsg_current_size = nlmsg_get_max_size(msg);
+	wpa_printf(MSG_DEBUG, "nl80211: current_size %zu new_size %zu",
+		   nlmsg_current_size, nlmsg_new_size);
+	nlmsg_free(msg);
+
+	if (nlmsg_new_size > nlmsg_current_size) {
+		wpa_printf(MSG_DEBUG, "nl80211: nlmsg_set_default_size to %zu",
+			   nlmsg_new_size);
+
+		nlmsg_set_default_size(nlmsg_new_size);
+	} else {
+		wpa_printf(MSG_DEBUG, "nl80211: skip nlmsg_set_default_size");
+	}
+
+	return 0;
+}
+
 static int nl80211_sta_steer(void *priv, const struct multi_ap_blacklist* entry, const u8 *bssid)
 {
 	int ret;
-- 
2.43.0

